runtime/gc_glue_java/HeadlessMixedObjectScanner.hpp

```diff
@@ -0,0 +1,191 @@
+
+#if !defined(HEADLESSMIXEDOBJECTSCANNER_HPP_)
+#define HEADLESSMIXEDOBJECTSCANNER_HPP_
+
+#include "j9.h"
+
+#include "ObjectScanner.hpp"
+
+class GC_HeadlessMixedObjectScanner : public GC_ObjectScanner
+{
+	/* Data Members */
+private:
+	fomrobject_t * const _endPtr;	/**< end scan pointer */
+	fomrobject_t *_mapPtr;			/**< pointer to first slot in current scan segment */
+	uintptr_t *_descriptionPtr;		/**< current description pointer */
+#if defined(J9VM_GC_LEAF_BITS)
+	uintptr_t *_leafPtr;			/**< current leaf description pointer */
+#endif /* J9VM_GC_LEAF_BITS */
+
+protected:
+
+public:
+
+	/* Member Functions */
+private:
+
+protected:
+	/**
+	 * @param env The scanning thread environment
+	 * @param[in] objectPtr the object to be processed
+	 * @param[in] flags Scanning context flags
+	 */
+	MMINLINE GC_HeadlessMixedObjectScanner(MM_EnvironmentBase *env, J9Class *clazzPtr, fomrobject_t *scanPtr, uintptr_t flags)
+		: GC_ObjectScanner(env, scanPtr, 0, flags)
+		, _endPtr((fomrobject_t *)((uint8_t*)_scanPtr + env->getExtensions()->mixedObjectModel.getSizeInBytesWithoutHeader(clazzPtr)))
+		, _mapPtr(_scanPtr)
+		, _descriptionPtr(NULL)
+#if defined(J9VM_GC_LEAF_BITS)
+		, _leafPtr(NULL)
+#endif /* J9VM_GC_LEAF_BITS */
+	{
+		_typeId = __FUNCTION__;
+	}
+
+	/**
+	 * Subclasses must call this method to set up the instance description bits and description pointer.
+	 * @param[in] env The scanning thread environment
+	 */
+	MMINLINE void
+	initialize(MM_EnvironmentBase *env, J9Class *clazzPtr)
+	{
+		GC_ObjectScanner::initialize(env);
+
+		if (J9_IS_J9CLASS_FLATTENED(clazzPtr)) {
+			fprintf(stderr, "!!! Found flattened class\n");
+		}
+
+		/* Initialize the slot map from description bits */
+		_scanMap = (uintptr_t)clazzPtr->instanceDescription;
+#if defined(J9VM_GC_LEAF_BITS)
+		_leafMap = (uintptr_t)clazzPtr->instanceLeafDescription;
+#endif /* J9VM_GC_LEAF_BITS */
+		if (_scanMap & 1) {
+			_scanMap >>= 1;
+			_descriptionPtr = NULL;
+#if defined(J9VM_GC_LEAF_BITS)
+			_leafMap >>= 1;
+			_leafPtr = NULL;
+#endif /* J9VM_GC_LEAF_BITS */
+			setNoMoreSlots();
+		} else {
+			_descriptionPtr = (uintptr_t *)_scanMap;
+			_scanMap = *_descriptionPtr;
+			_descriptionPtr += 1;
+#if defined(J9VM_GC_LEAF_BITS)
+			_leafPtr = (uintptr_t *)_leafMap;
+			_leafMap = *_leafPtr;
+			_leafPtr += 1;
+#endif /* J9VM_GC_LEAF_BITS */
+		}
+	}
+
+public:
+	/**
+	 * In-place instantiation and initialization for mixed object scanner.
+	 * @param[in] env The scanning thread environment
+	 * @param[in] objectPtr The object to scan
+	 * @param[in] allocSpace Pointer to space for in-place instantiation (at least sizeof(GC_MixedObjectScanner) bytes)
+	 * @param[in] flags Scanning context flags
+	 * @return Pointer to GC_MixedObjectScanner instance in allocSpace
+	 */
+	MMINLINE static GC_HeadlessMixedObjectScanner *
+	newInstance(MM_EnvironmentBase *env, J9Class* clazzPtr, fomrobject_t* scanPtr, void *allocSpace, uintptr_t flags)
```

Comment:
Do you need this method at all? It is just parent class for `GC_MixedObjectScanner` so no need there. For new scanner for Flattened Arrays it would be built in so should be better way to initialize/re-initialize it is statically (for it) allocated memory